### PR TITLE
[Model] Implement Neural Collaborative Filtering with MXNet

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
 Implement Neural Collaborative Filtering with MXNet according to mlperf v0.5 configuration.

Evaluation accuracy on MovieLens 20m dataset:

|dtype|HR@10|NDCG@10|
|:---:|:--:|:--:|
|float32|0.6393|0.3849|
|int8|0.6366|0.3824|

@pengzhao-intel @TaoLv @zhreshold

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] data preprocess for MovieLens datasets
- [x] ncf training on MovieLens dataset
- [x] ncf evaluation and benchmark on MovieLens dataset
- [x] ncf int8 quantization on MovieLens dataset
- [x] ncf training/inference/quantization smoke test

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@wkcn could you help take a review in case you're interested in?

**Comment by wkcn** [comment]
@pengzhao-intel Sorry that I am not familiar with neural collaborative filtering, but I try to learn it : )

**Comment by pengzhao-intel** [comment]
@xinyu-intel is this work done and ready to merge?

**Comment by xinyu-intel** [comment]
yes:)

**Comment by wkcn** [comment]
Thank @xinyu-intel for the great contribution!
Sorry that I am busy recently, and I am taking a review on this PR today.

File: example/neural_collaborative_filtering/core/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    user = []
+    item = []
+    score = []
+    with open(filename, 'r') as f:
+        num_samples = 0
+        for line in f:
+            tks = line.strip().split('\t')
+            if len(tks) != 3:
+                continue
+            if (int(tks[0]) > 138000 or int(tks[1]) > 26700):
```
Comment (wkcn) [inline example/neural_collaborative_filtering/core/evaluate.py]: Why are the samples removed?

File: example/neural_collaborative_filtering/core/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    user = mx.nd.array(users, dtype='int32').as_in_context(ctx)
+    item = mx.nd.array(items, dtype='int32').as_in_context(ctx)
+    label = mx.nd.zeros(len(user)).as_in_context(ctx)
+    data = {'user': user, 'item': item}
+    label = {'softmax_label':label}
+    eval_iter = mx.io.NDArrayIter(data=data, label=label, batch_size=batch_size)
+    preds = []
+    for batch in eval_iter:
+        model.forward(batch)
+        mx.nd.waitall()
```
Comment (wkcn) [inline example/neural_collaborative_filtering/core/evaluate.py]: waitall is not necessary since asnumpy will call wait_to_read.

File: example/neural_collaborative_filtering/core/dataset.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+class NCFTestData(object):
+    def __init__(self, path):
+        '''
+        Constructor
+        path: converted data root
+        testRatings: converted test ratings data
+        testNegatives: negative samples for evaluation dataset
+        '''
+        self.testRatings = self.load_rating_file_as_list(path + "/test-ratings.csv")
```
[the remark this answers is not in the document]
Comment (xinyu-intel) [reply example/neural_collaborative_filtering/core/dataset.py]: thx, will fix later:)

File: example/neural_collaborative_filtering/core/dataset.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    def __getitem__(self, idx):
+        if idx % (self.nb_neg + 1) == 0:
+            idx = idx // (self.nb_neg + 1)
+            return self.data[idx][0], self.data[idx][1], np.ones(1, dtype=np.float32).item()  # noqa: E501
+        else:
+            idx = idx // (self.nb_neg + 1)
+            u = self.data[idx][0]
+            j = mx.random.randint(0, self.nb_items).asnumpy().item()
+            while (u, j) in self.mat:
+                j = mx.random.randint(0, self.nb_items).asnumpy().item()
```
[the remark this answers is not in the document]
Comment (xinyu-intel) [reply example/neural_collaborative_filtering/core/dataset.py]: this function is used to random choose a negative item which user isn't interested in. ref: https://github.com/mlperf/training/blob/948db9b11cdfa7d953769e53c560396f41617f1b/recommendation/pytorch/dataset.py#L44

File: example/neural_collaborative_filtering/core/evaluate.py (hunk 1 above)
Comment (xinyu-intel) [reply example/neural_collaborative_filtering/core/evaluate.py]: thx, it's an old implement only for benchmark and calibration now. will remove this line later.

File: example/neural_collaborative_filtering/core/evaluate.py (hunk 2 above)
Comment (xinyu-intel) [reply example/neural_collaborative_filtering/core/evaluate.py]: make sense:)
